skyportal/models.py

```diff
@@ -718,99 +718,107 @@ class Obj(Base, ha.Point):
+    @hybrid_method
+    def last_detected_mag(self, user):
         """Magnitude at which the object was last detected above a given S/N (3.0 by default)."""
         detections = [
             (phot.iso, phot.mag)
-            for phot in self.photometry
+            for phot in Photometry.query_records_accessible_by(user)
+            .filter(Photometry.obj_id == self.id)
+            .all()
             if phot.snr is not None and phot.snr > PHOT_DETECTION_THRESHOLD
         ]
         return max(detections, key=(lambda x: x[0]))[1] if detections else None
 
     @last_detected_mag.expression
-    def last_detected_mag(cls):
+    def last_detected_mag(cls, user):
         """Magnitude at which the object was last detected above a given S/N (3.0 by default)."""
-        last_detected = (
-            sa.select([cls.id, sa.func.max(Photometry.mjd).label("max_mjd")])
-            .where(Photometry.obj_id == cls.id)
-            .where(Photometry.snr.isnot(None))
-            .where(Photometry.snr > PHOT_DETECTION_THRESHOLD)
-            .group_by(cls.id)
-            .alias()
-        )
         return (
-            sa.select([Photometry.mag])
-            .where(Photometry.obj_id == cls.id)
-            .where(Photometry.snr.isnot(None))
-            .where(Photometry.snr > PHOT_DETECTION_THRESHOLD)
-            .where(cls.id == last_detected.c.id)
-            .where(Photometry.mjd == last_detected.c.max_mjd)
+            Photometry.query_records_accessible_by(
+                user, columns=[Photometry.mag], mode="read"
+            )
+            .filter(Photometry.obj_id == cls.id)
+            .filter(Photometry.snr.isnot(None))
+            .filter(Photometry.snr > PHOT_DETECTION_THRESHOLD)
+            .order_by(Photometry.mjd.desc())
+            .limit(1)
             .label('last_detected_mag')
         )
 
-    @hybrid_property
-    def peak_detected_at(self):
+    @hybrid_method
+    def peak_detected_at(self, user):
         """UTC ISO date at which the object was detected at peak magnitude above a given S/N (3.0 by default)."""
         detections = [
             (phot.iso, phot.mag)
-            for phot in self.photometry
+            for phot in Photometry.query_records_accessible_by(user)
+            .filter(Photometry.obj_id == self.id)
+            .all()
             if phot.snr is not None and phot.snr > PHOT_DETECTION_THRESHOLD
         ]
         return max(detections, key=(lambda x: x[1]))[0] if detections else None
 
     @peak_detected_at.expression
-    def peak_detected_at(cls):
+    def peak_detected_at(cls, user):
         """UTC ISO date at which the object was detected at peak magnitude above a given S/N (3.0 by default)."""
         return (
-            sa.select([Photometry.iso])
-            .where(Photometry.obj_id == cls.id)
-            .where(Photometry.snr.isnot(None))
-            .where(Photometry.snr > PHOT_DETECTION_THRESHOLD)
+            Photometry.query_records_accessible_by(
```

Comment:
does it matter that this used to return a `select` instance and now returns a `query` instance? not sure if they're interchangeable, so just making sure

Reply:
No I don't think it matters. I didn't look explicitly for it in the SQLAlchemy docs, but all the tests that use these passed locally for me. We have several tests in test_sources.py for things like filtering on peak mag and latest detection which should use these expressions.

Reply:
nice